python/ray/air/session.py

```diff
@@ -0,0 +1,233 @@
+    def type(self) -> SessionType:
+        """The type of the session.
+
+        Returns:
+            The type of the session.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def report(self, metrics: Dict, checkpoint: Optional[Checkpoint] = None) -> None:
+        """Report metrics and optionally save checkpoint.
+
+        The checkpoint is guaranteed to be saved onto Driver node.
+
+        Each invocation of this method will automatically increment the underlying
+        iteration number. The physical meaning of this "iteration" is defined by
+        user (or more specifically the way they call ``report``).
+        It does not necessarily map to one epoch.
+
+        This API is supposed to replace the legacy ``tune.report``,
+        ``with tune.checkpoint_dir``, ``train.report`` and ``train.save_checkpoint``.
+        Please avoid mixing them together.
+
+        There is no requirement on what is the underlying representation of the checkpoint.
+        All forms are accepted and (will be) handled by AIR in an efficient way.
+
+        Example:
+            .. code-block: python
+
+                from ray.air.session import get_session
+                from ray.air.checkpoint import Checkpoint
+                ########## Using it in the *per worker* train loop (TrainSession) ###################
+                def train_func():
+                    session = get_session()
+                    model = build_model()
+                    # ...
+
+                    # TEMPORARY!!
+                    # This is bad if user has to do this. Also we lose track of the ownership of the
+                    # temporary directory and when can clean it up.
+                    # Can be avoided, if we expose session.local_dir for one to write file to upfront.
+                    # So that multiple workers on the same physical machine won't step onto each other.
+                    # This can also avoid unnecessary copying and make bookkeeping and checkpoint deleting
+                    # easier.
+                    import tempfile
+                    temp_dir = tempfile.TemporaryDirectory()
+                    model.save(temp_dir.name, overwrite=True)
+                    session.report(
+                        metrics={"foo": "bar"}, checkpoint=Checkpoint.from_directory(temp_dir.name)
+                    )
+                scaling_config = {"num_workers": 2}
+                trainer = TensorflowTrainer(
+                    train_loop_per_worker=train_func, scaling_config=scaling_config
+                )
+                result = trainer.fit()
+                # If you navigate to result.checkpoint's path, you will find the content of ``model.save()``
+                # under it.
+
+                # Aspirational - cloud storage story
+                def train_func():
+                    session = get_session()
+                    # TBD
+                    uri = session.remote_dir
```

Comment:
Just want to call out that this may be controversial.

Currently, tune handles all cloud checkpointing stuff behind the scene, user just needs to do 
```
with tune.checkpoint_dir(epoch) as checkpoint_dir:
  ....
```
And things will be automatically synced to cloud if `sync_config` is configured.

Now we are asking user to do `checkpoint.to_uri` themselves and checkpoint this checkpoint to AIR.

This is all because we are pushing Checkpoint to be the main abstraction.

Reply:
Thinking more, we can actually have all that behind the scene.
From user's perspective, they just do
def train_func():
  session = get_session()
  checkpoint.to_directory(session.local_dir)
  session.report({"foo": "bar", checkpoint=checkpoint})  # Session will be responsible for uploading to cloud if applicable

Behind the scene,
class Session:
  self._remote_dir = None  # configured according to SyncConfig

  def report(...):
    ...
    if self._remote_dir:
      # do upload to cloud

 @krfricke wdyt?

Reply:
I really don't think this belongs in user code--- why can't it be entirely in SyncConfig?

Reply:
Updated doc strings, ptal.